api/python/flaschen.py

```diff
@@ -81,3 +99,36 @@ def set(self, x, y, color):
   def send(self):
     '''Send the updated pixels to the display.'''
     self._sock.send(self._data)
+
+  def send_array(self, pixels, offset):
+    # (numpy.typing.ArrayLike, tuple[int, int, int]) -> None
+    '''Send an array of pixels to the given offset.
+
+    Can be used as an alternative to the send method.
+    The initial transparent option is respected.
+    The initial width, height, and layer is ignored.
+
+    Args:
+      pixels: An array-like of RGB pixels, shaped as (height, width, RGB)
+              Color values are expected to be from 0 to 255.
+      offset: The (x, y, layer) offset.
+              X and Y begin at the top-left pixel.
+    '''
+    import numpy as np
+    array = np.asarray(pixels, dtype=np.uint8)
+    if len(array.shape) != 3 or array.shape[2] != 3:
+      raise TypeError(
+        "Pixel array must be shape (height, width, 3), got %r" % (array.shape,)
+      )
+    if not self.transparent:
+      array = array.copy(order="C")  # Prevent modifying the original array.
+      array[(array == (0, 0, 0)).all(axis=-1)] = (1, 1, 1)
+
+    header = _HEADER_P6_FT % {
+      b"width": array.shape[1],
+      b"height": array.shape[0],
+      b"x": offset[0],
+      b"y": offset[1],
+      b"z": offset[2],
+    }
+    self._sock.send(header + array.tobytes())
```

Comment:
one of the reasons to use the `_data` member was that there is no expensive allocation and conversion needed at runtime. Here, you have to do that at runtime and copy arrays together. I don't know or use Python so maybe this is acceptable, as performance is not necessarily a goal. But if there is a `writev()` in Python, maybe use that to at least avoid having to copy both arrays into one just to call `send()`.

Reply:
Performance is relative. In the `Flaschen.set` method the cost to convert the Python types to insert into `_data` is incredibly expensive for something that's supposed to be called on every changed pixel.  Even in Numpy setting one pixel by itself is slow.

This operation has allocations but not many of them.  Python isn't low level enough to where removing all allocations is an option.  Like the objects used to manipulate `_data` are still allocated.

I can profile this code if you want proof of its performance.  I'd like to suggest that until a profiler says otherwise, this method is efficient enough, even with the allocations.